### PR TITLE
Connector instance apis crud

### PR DESCRIPTION
**Pull request by LakshmiSagarNS**

### CODE REVIEW

File: api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreateValidationSchema.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            "properties": {
+                
+                "dataset_id":{
+                    "type":"string"
+
+                },
+                "connector_id":{
+                    "type":"string"
+                
+                }
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreateValidationSchema.json:39]: What about the other fields in the request, not just minimal we have to support all the fields with minimal fields as required

File: api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import _ from "lodash";
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts:1]: File name should start with capital C only

File: api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { Request, Response } from "express";
+import httpStatus from "http-status";
+import { schemaValidation } from "../../services/ValidationService";
+import { ResponseHandler } from "../../helpers/ResponseHandler";
+import ConnectorInstanceCreate from "./ConnectorInstanceCreateValidationSchema.json";
+import { obsrvError } from "../../types/ObsrvError";
+import { ConnectorInstances } from "../../models/ConnectorInstances";
+import { connectorInstance } from "../../services/ConnectorInstanceService";
+
+// validating the Schema using the connectorInstanceCreateValidationSchema.json present in the ConnectorInstanceCreate folder.
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts:11]: No need of the comments

File: api-service/src/services/ConnectorInstanceService.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { ConnectorInstances } from "../models/ConnectorInstances";
```
Comment (HarishGangula) [inline api-service/src/services/ConnectorInstanceService.ts:1]: Check the dataset implementation service file should not request and response dependency

File: api-service/src/routes/Router.ts [hunk 5]
```diff
@@ -55,5 +57,11 @@ router.post("/connectors/list", setDataToRequestObject("api.connectors.list"), o
 router.get("/connectors/read/:id", setDataToRequestObject("api.connectors.read"), onRequest({entity: Entity.Management }), ConnectorsRead);
 router.post("/datasets/import", setDataToRequestObject("api.datasets.import"), onRequest({ entity: Entity.Management }), DatasetImport);
 
+router.post("/connector-instances/create",setDataToRequestObject("api.connector.instance.create"),onRequest({ entity: Entity.Management }),connectorInstanceCreate);
+//router.get("connector-instances/read/:id",setDataToRequestObject("api.connector.instance.get"),onRequest({ entity: Entity.Management }),connectorInstanceGet);
```
Comment (HarishGangula) [inline api-service/src/routes/Router.ts:61]: Remove the commented code

File: api-service/src/routes/Router.ts [hunk 6]
```diff
@@ -55,5 +57,11 @@ router.post("/connectors/list", setDataToRequestObject("api.connectors.list"), o
 router.get("/connectors/read/:id", setDataToRequestObject("api.connectors.read"), onRequest({entity: Entity.Management }), ConnectorsRead);
 router.post("/datasets/import", setDataToRequestObject("api.datasets.import"), onRequest({ entity: Entity.Management }), DatasetImport);
 
+router.post("/connector-instances/create",setDataToRequestObject("api.connector.instance.create"),onRequest({ entity: Entity.Management }),connectorInstanceCreate);
```
Comment (HarishGangula) [inline api-service/src/routes/Router.ts:60]: Following the indentation and linting

File: api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+const validateRequest = async (req: Request) => {
+
+    const isRequestValid: Record<string, any> = schemaValidation(req.body,ConnectorInstanceCreate)
+    if (!isRequestValid.isValid) {
+        throw obsrvError("", "CONNECTOR_INSTANCE_INVALID_INPUT", isRequestValid.message, "BAD_REQUEST", 400)
+    }
+    // checking whether the connector_instance exist using the id with the help of get and findByPk method.
+
+    const connectorInstanceId = _.get(req, ["body","request","id"])
+    const isConnectorInstanceIdExists = await ConnectorInstances.findByPk(connectorInstanceId);
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts:22]: No direct interaction with db models in the controller

**Review by HarishGangula — CHANGES_REQUESTED** [review]
Fix all the review comments and understand the semantic meaning of the separation of the concerns from dataset APIs as explained

File: api-service/src/routes/Router.ts (hunk 6 above)
Comment (LakshmiSagarNS) [reply api-service/src/routes/Router.ts]: changes done.

File: api-service/src/routes/Router.ts (hunk 5 above)
Comment (LakshmiSagarNS) [reply api-service/src/routes/Router.ts]: removed

File: api-service/src/services/ConnectorInstanceService.ts (hunk 4 above)
Comment (LakshmiSagarNS) [reply api-service/src/services/ConnectorInstanceService.ts:1]: variables name are changed

File: api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts (hunk 3 above)
Comment (LakshmiSagarNS) [reply api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts]: comments are removed

File: api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts (hunk 2 above)
Comment (LakshmiSagarNS) [reply api-service/src/controllers/ConnectorInstanceCreate/connectorInstanceCreate.ts]: change done

File: api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreateValidationSchema.json (hunk 1 above)
Comment (LakshmiSagarNS) [reply api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreateValidationSchema.json]: i have updated the following required fields.
data_format and connector_config

File: api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreate.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+             throw obsrvError(connectorInstanceId, "CONNECTOR_INSTANCE_EXISTS", `ConnectorInstance Already exists with id:${connectorInstanceId}`, "CONFLICT", 409)
+        }
+
+}
+
+const connectorInstanceCreate = async (req: Request, res: Response) => {
+    validateRequest(req)
+    const dataset_id = _.get(req, ["body", "request", "dataset_id"]);
+    const connector_id = _.get(req, ["body", "request", "connector_id"]);
+    const id = `${connector_id}.${dataset_id}`;
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceCreate/ConnectorInstanceCreate.ts:30]: handing the trailing space issues

File: api-service/src/controllers/ConnectorInstanceRead/ConnectorInstanceRead.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import { ResponseHandler } from "../../helpers/ResponseHandler";
+import { connectorInstance } from "../../services/ConnectorInstanceService";
+import { obsrvError } from "../../types/ObsrvError";
+
+const defaultFields = ["connector_id", "dataset_id", "status", "data_format", "operations_config", "connector_stats", "connector_state", "created_by", "updated_by", "created_date", "updated_date", "published_date"]
+
+const validateRequest = (req: Request) => {
+    const { id } = req.params;
+    const { fields } = req.query;
+    if (fields && typeof fields !== 'string') {
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceRead/ConnectorInstanceRead.ts:13]: make the validation simple with lodash methods

File: api-service/src/controllers/ConnectorInstanceUpdate/ConnectorInstanceUpdate.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    const connectorInstanceBody = req.body.request
+    const { id, ...rest } = connectorInstanceBody
+    if (_.isEmpty(rest)) {
+        throw obsrvError(connector_id, "CONNECTOR_INSTANCE_UPDATE_NO_FIELDS", "Provide atleast one field in addition to the id to update the connector instance", "BAD_REQUEST", 400)
+    }
+
+    const connectorInstanceExist = await connectorInstance.checkConnectorInstanceExists(connector_id)
+    if (!connectorInstanceExist) {
+        throw obsrvError(connector_id, "CONNECTOR_INSTANCE_DOES_NOT_EXIST", "Provide the valid connector instance id", "BAD_REQUEST", 400)
```
Comment (HarishGangula) [inline api-service/src/controllers/ConnectorInstanceUpdate/ConnectorInstanceUpdate.ts:26]: 404 status code

**Review by HarishGangula — CHANGES_REQUESTED** [review]

**Pull request closed without merging**